Use token auth when widgetize

flamisz opened the pull request:
### Description:

fixes #17335 

### Review

* [ ] Functional review done
* [ ] Potential edge cases thought about (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] Usability review done (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] Security review done [see checklist](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] Code review done
* [ ] Tests were added if useful/possible
* [ ] Reviewed for breaking changes
* [ ] Developer changelog updated if needed
* [ ] Documentation added if needed
* [ ] Existing documentation updated if needed

flamisz committed:
use token auth when widgetize

tsteur commented on core/FrontController.php:
```diff
@@ -428,8 +428,13 @@ public function init()
         if (Common::getRequestVar('token_auth', '', 'string') !== ''
             && Request::shouldReloadAuthUsingTokenAuth(null)
         ) {
+            if (!Common::isPhpCliMode()) {
```
@flamisz might be worth trying not to use the session authenticator when a token_auth is used in the query. Eg return `null` in `makeSessionAuthenticator()` if a session is set? Unsetting super user access etc may have side effects not sure

flamisz committed:
Update FrontController.php

flamisz commented:
Hi @tsteur, I updated the PR, not resetting the super user access. During this update I've seen this comment in the code:

https://github.com/matomo-org/matomo/blob/104135aebddd112567f5ec17b8f11e757c7e57f2/core/FrontController.php#L426-L433

But I modified the `reloadAccess` method and now it works with widgets as well. Do you think it was delibaretly did this way? Could it be any side effect somewhere?
https://github.com/matomo-org/matomo/blob/104135aebddd112567f5ec17b8f11e757c7e57f2/core/Access.php#L137

tsteur commented:
@flamisz I think we can undo the changes in `Access.php` and apply this change:

```diff
--- a/core/FrontController.php
+++ b/core/FrontController.php
@@ -675,7 +675,8 @@ class FrontController extends Singleton
             return null;
         }
 
-        if (Common::getRequestVar('token_auth', '', 'string') !== '') {
+        if (Common::getRequestVar('token_auth', '', 'string') !== ''
+            && !Common::getRequestVar('force_api_session')) {
             return null;
         }
 ```

It basically means if there's a `force_api_session=1` parameter then we still load the session and authenticate the session first which is needed for the session token to work. This way the dashboard etc will work and all other requests should be working too. It be only an issue if users falsely have `force_api_session=1` set in the request and don't use a session token but a real token auth. Then it won't work but then it generally shouldn't work anyway AFAIK.

It be still good for someone to review this though additionally.

flamisz commented:
> @flamisz I think we can undo the changes in `Access.php` and apply this change:

I just realized the way it was is not working and tried to figure out how it should 😄. Thanks, I was going to check the `force_api_session` but wasn't sure where yet. 

I'll test it how it works this way.

flamisz committed:
check if force_api_session set

flamisz committed:
fix test error message

sgiehl commented on core/FrontController.php:
```diff
@@ -430,6 +430,7 @@ public function init()
         ) {
             Request::reloadAuthUsingTokenAuth();
             Request::checkTokenAuthIsNotLimited($module, $action);
+            Session::close();
```
Any reason for closing the session at this point?

flamisz replied:
During the process somewhere we start it, but actually we don't need it as we handle this request as stateless, so I thought we can close it (the original idea was coming from @tsteur).

sgiehl replied:
In theory that might make sense, but not sure if that could cause any regressions if someone tries to perform a controller action with a token_auth, that actually tries to change the session. But maybe we can simply change it and wait if any regressions are reported.

tsteur replied:
It should be fine to close the session if `force_api_session` is not present in the URL AFAIK. There could be a regression but it would be hard to know. In theory though in these cases the session shouldn't be started anyway.

diosmosis replied:
Would this be an issue for the "embed all of matomo" use case (one big user uses this, can't remember their name starts w/ an 'M')? Ie, the `[General] enable_framed_pages` `[General] enable_framed_settings` config. I would think the session might be written to there.

diosmosis replied:
Also, thought it might be useful to have test pages for these, so created: https://github.com/matomo-org/test-examples/pull/6

flamisz replied:
I'm not familiar with those settings (framed). I tried with the new `test-example` and worked. How else should we test it? @diosmosis @tsteur

diosmosis replied:
They allow users to embed whole matomo pages. One company I know of uses it to apply their own branding to all of Matomo and embed it as part of their website for their customers.

I think to test that it would work we'd need to write the session when a page is framed and see if it persists? Maybe sending a notification from PHP and seeing it in the embedded admin page?

flamisz replied:
I tried something.

1. created a persistent notification in a controller
2. uncommented the line where I created it
3. reloaded Matomo page, the notification was there

I did the same when Matomo was running inside an iframe. This time at step 3, the notification was not there. It means the session was not persisted anyway.

What is the conclusion here? Should we keep that `Session::close()` line?
ping @tsteur @diosmosis @sgiehl

diosmosis replied:
We could also just do the Session::close() if module==Widgetize. That might be enough?

flamisz replied:
done 👍

tsteur replied:
I realise using `new SessionNamespace` will actually lounge a session. Eg in `redirectDashboardToWelcomePage` method it might start a session and that might even happen when it's an API call or when the action is embedded as an iframe. Closing it now could break some things that were working before (whether they should have been working before or not is a different question). I didn't debug but likely the close shouldn't do anything because ideally it doesn't start the session in the first place so I'm tempted to simply remove the close call. And eg if `force_api_session=1` then the session might be still needed in order to write session changes

tsteur replied:
If the request is an API call then we for sure don't want to use any session but widgetized requests may use a session I would say

diosmosis replied:
@flamisz I think ^ means my idea wouldn't be enough

flamisz replied:
Ok, I tried a couple of things, tested what I could, and looks like the `force_api_session` does the job here. If that's false, we don't have a session, so don't need to close. 

I couldn't find a case when we really need to close it.

flamisz committed:
close session only when widgetize

flamisz committed:
no session close

diosmosis committed:
fix test
